### PR TITLE
Add cert validation option to requestheader

### PR DESCRIPTION
**Pull request by liggitt**
@brenton fyi

### CODE REVIEW

**Comment by liggitt** [comment]
@deads2k

File: pkg/auth/authenticator/request/x509request/x509.go [hunk 1]
```diff
@@ -60,7 +61,35 @@ func (a *Authenticator) AuthenticateRequest(req *http.Request) (user.Info, bool,
 			}
 		}
 	}
-	return nil, false, errors.NewAggregate(errlist)
+	return nil, false, kerrors.NewAggregate(errlist)
+}
+
+// Verifier implements request.Authenticator by extracting user info from verified client certificates
```
Comment (deads2k) [inline pkg/auth/authenticator/request/x509request/x509.go]: Clarify. I think this is wrapping a RequestAuthenticator inside of another RequestAuthenticator that requires a valid client cert as opposed to pulling the user info from a client cert.

File: pkg/cmd/server/origin/auth.go [hunk 2]
```diff
@@ -552,6 +557,21 @@ func (c *AuthConfig) getAuthenticationRequestHandlerFromType(authRequestHandlerT
 			UserNameHeaders: c.RequestHeaders,
 		}
 		authRequestHandler = headerrequest.NewAuthenticator(authRequestConfig, identityMapper)
+
+		// Wrap with an x509 verifier
+		if len(c.RequestHeaderCAFile) > 0 {
+			caData, err := ioutil.ReadFile(c.RequestHeaderCAFile)
+			if err != nil {
+				glog.Fatalf("Error reading %s: %v", c.RequestHeaderCAFile, err)
```
Comment (deads2k) [inline pkg/cmd/server/origin/auth.go]: TODO: remove the fatals from this method.

File: pkg/cmd/server/origin_master.go [hunk 3]
```diff
@@ -210,21 +217,84 @@ func (cfg Config) newCA() (*crypto.CA, error) {
...
+	}
+	roots := x509.NewCertPool()
+	for _, root := range certs {
+		roots.AddCert(root)
+	}
+	return roots, nil
+}
+
+func (cfg Config) getOAuthClientCertCAs() ([]*x509.Certificate, error) {
+	auth, err := cfg.BuildAuthConfig()
```
Comment (deads2k) [inline pkg/cmd/server/origin_master.go]: This feels cyclical.  Building the master config requires building the authconfig?  Any way to pull out the specific parts we care about into config.Get* methods?

Comment (liggitt) [inline pkg/cmd/server/origin_master.go]: Pulled out the `OPENSHIFT_OAUTH_REQUEST_HEADER_CA_FILE` lookup into `GetOAuthRequestHeaderCAFile`

File: pkg/cmd/server/origin/auth.go [hunk 4]
```diff
@@ -229,7 +234,10 @@ func (c *AuthConfig) InstallAPI(container *restful.Container) []string {
 
 	oauthEtcd := oauthetcd.New(c.EtcdHelper)
 
-	authRequestHandler, authHandler, authFinalizer := c.getAuthorizeAuthenticationHandlers(mux)
+	authRequestHandler, authHandler, authFinalizer, err := c.getAuthorizeAuthenticationHandlers(mux)
+	if err != nil {
+		glog.Fatal(err)
```
Comment (liggitt) [inline pkg/cmd/server/origin/auth.go]: This is the unified glog.Fatal for now

**Comment by liggitt** [comment]
@deads2k comments addressed

**Comment by deads2k** [comment]
lgtm [merge]

**Pull request merged**